core/src/test/java/apoc/refactor/GraphRefactoringTest.java

```diff
@@ -994,5 +998,148 @@ public void testRefactorCategorizeNoDups() {
         assertEquals(8, relsCount);
         db.executeTransactionally("DROP CONSTRAINT ON (n:`" + label + "`) ASSERT n.`" + targetKey + "` IS UNIQUE");
     }
+
+    @Test
+    public void testMergeNodeShouldNotCreateSelfRelationshipsInPreExistingSelfRel() {
+        db.executeTransactionally("CREATE (a:TestNode {a:'a'})-[:TEST_REL]->(b:TestNode {a:'b'})-[:TEST_REL]->(c:TestNode {a:'c'})\n" +
+                "WITH a, c CREATE (a)-[:TEST_REL]->(a) WITH c CREATE (c)-[:TEST_REL]->(c);");
+        testCall(db, "MATCH (n:TestNode) WITH collect(n) as nodes CALL apoc.refactor.mergeNodes(nodes, {mergeRels: true, selfRel: false}) yield node return node",
+                (r) -> {
+                    Node node = (Node) r.get("node");
+                    assertFalse(node.getRelationships().iterator().hasNext());
+                });
+    }
+
+    @Test
+    public void testMergeNodeShouldCreateSelfRelationshipsInPreExistingSelfRel() {
+        db.executeTransactionally("CREATE (a:TestNode {a:'a'})-[:TEST_REL]->(b:TestNode {a:'b'})-[:TEST_REL]->(c:TestNode {a:'c'})\n" +
+                "WITH a, c CREATE (a)-[:TEST_REL]->(a) WITH c CREATE (c)-[:TEST_REL]->(c);");
+        testCall(db, "MATCH (n:TestNode) WITH collect(n) as nodes CALL apoc.refactor.mergeNodes(nodes, {mergeRels: true, selfRel: true}) yield node return node",
+                (r) -> {
+                    Node node = (Node) r.get("node");
+                    Iterator<Relationship> relIterator = node.getRelationships().iterator();
+                    String relName = relIterator.next().getType().name();
+                    assertEquals("TEST_REL", relName);
+                    assertFalse(relIterator.hasNext());
+                });
+    }
+
+    @Test
+    public void testMergeNodeShouldNotCancelOtherRelsWithSelfRelsTrue() {
+        db.executeTransactionally("CREATE (a:A {a:'a'})-[:KNOWS]->(b:B {a:'b'})-[:KNOWS]->(c:C {a:'c'})\n" +
+                "WITH a,b,c CREATE (a)-[:KNOWS]->(a) WITH a,b,c CREATE (a)-[:KNOWS]->(c) WITH c,b CREATE (c)-[:KNOWS]->(b);");
+        testCall(db, "MATCH (n:A), (m:B) WITH [n,m] as nodes CALL apoc.refactor.mergeNodes(nodes, {mergeRels: true, selfRel: true}) yield node return node",
+                (r) -> {
+                    Node node = (Node) r.get("node");
+                    List<Relationship> relationships = IteratorUtils.toList(node.getRelationships().iterator());
+
+                    assertEquals(3, relationships.size());
```

Comment:
same remark as above: shouldn't there be 4 relationship (as A to A should be preserved by default)?